src/uct/cuda/cuda_copy/cuda_copy_md.c

```diff
@@ -290,52 +290,128 @@ static size_t uct_cuda_copy_md_get_total_device_mem(CUdevice cuda_device)
     return 1; /* return 1 byte to avoid division by zero */
 }
 
+static ucs_status_t
+uct_cuda_copy_detect_vmm(const void *address, CUmemorytype *cuda_mem_type,
+                         unsigned *is_vmm_allocation)
+{
+    ucs_status_t status = UCS_OK;
+#if HAVE_CUDA_FABRIC
+    CUmemAllocationProp prop = {};
+    CUmemGenericAllocationHandle alloc_handle;
+
+    /* Check if memory is allocated using VMM API and see if host memory needs
+     * to be treated as pinned device memory */
+    status = UCT_CUDADRV_FUNC_LOG_DEBUG(
+            cuMemRetainAllocationHandle(&alloc_handle, (void*)address));
+    if (status != UCS_OK) {
+        status = UCS_OK;
+        goto ret;
+    }
+
+    *is_vmm_allocation = 1;
+    if (*cuda_mem_type != CU_MEMORYTYPE_HOST) {
+        goto mem_release;
+    }
+
+    status = UCT_CUDADRV_FUNC_LOG_DEBUG(
+            cuMemGetAllocationPropertiesFromHandle(&prop, alloc_handle));
+    if (status != UCS_OK) {
+        status = UCS_ERR_INVALID_ADDR;
+        goto mem_release;
+    }
+
+    /* if location type is HOST the memory can be potentially accessed by CEs */
+    if ((prop.location.type == CU_MEM_LOCATION_TYPE_HOST) ||
+        (prop.location.type == CU_MEM_LOCATION_TYPE_HOST_NUMA) ||
+        (prop.location.type == CU_MEM_LOCATION_TYPE_HOST_NUMA_CURRENT)) {
+        *cuda_mem_type = CU_MEMORYTYPE_DEVICE;
+    } else {
+        status = UCS_ERR_INVALID_ADDR;
+    }
+
+mem_release:
+    status = UCT_CUDADRV_FUNC_LOG_DEBUG(cuMemRelease(alloc_handle));
+ret:
+#endif
+    return status;
+}
+
+static void
+uct_cuda_copy_set_sync_memops(const void *address, unsigned is_vmm_allocation)
+{
+    unsigned value = 1;
+    CUresult cu_err;
+
+#if HAVE_CUDA_FABRIC
+    if (is_vmm_allocation) {
+        cu_err = cuCtxSetFlags(CU_CTX_SYNC_MEMOPS);
+        if (cu_err != CUDA_SUCCESS) {
+            ucs_warn("cuCtxSetFlags(CU_CTX_SYNC_MEMOPS) for %p error: %s",
+                     address, uct_cuda_base_cu_get_error_string(cu_err));
+        }
```

Comment:
maybe use `UCT_CUDADRV_FUNC` here?

Reply:
why do we set sync flag for the context for vmm only?

Reply:
@brminich VMM memory cannot use SYNC_MEMOPS attribute and the mechanism exposed is the above API.

Reply:
ok, what is about using `UCT_CUDADRV_FUNC` here?